### DISCUSSION TITLE
Differences among auxiliary kernel, material and userobject.

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#21956] [Q&A General] [answered]
Hello buddy,

It has been awhile for me to undrstand the differences among auxiliary kernel, material and userobject.
Anyone can give a brief introduction for the differences among these three cases.
For me, they all have the common chracteristic, that is, compute something then other objects can use their results.

Thank you in advance.

Best regards,
Qiao

**Comment by GiudGiud** [comment, marked as answer]
Docs pages should help
https://mooseframework.inl.gov/syntax/Materials/index.html
https://mooseframework.inl.gov/syntax/AuxKernels/index.html
https://mooseframework.inl.gov/syntax/UserObjects/index.html

in short, you should try to use material properties if you can, especially if using AD.
auxkernels are equally desirable to compute fields, but dont support AD, and cost more memory (storing an extra field)

If your calculation is very complicated, then you ll want to make a custom user object.
